### PR TITLE
End2end test for metric tank in docker

### PR DESCRIPTION
**Pull request by OOM-Killer**
This Pull Request is related to #346. I cleaned up the commit history, added more comments, and fixed everything as requested in the comments on #346. Apart from the requested changes the content should be the same.

Quote from #346:

The purpose of this test is to run the whole docker-compose environment, send data through it, and see if that data arrives.

The end2end test uses `fakemetrics` to generate `1000` metric points per second. After waiting for a while it uses the `verify_metrics_receive.py` script to query Graphite how many metrics Metrictank has received during the last `10` seconds. If at average it has received at least `1000` per second the test passes.

### CODE REVIEW

File: scripts/wait_for_endpoint.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/bin/sh
+
+
+WAIT_TIMEOUT=${WAIT_TIMEOUT:-10}
+
+# test if we're using busybox for timeout
+timeout_exec=$(readlink `which timeout` | awk -F '/' '{print $NF}')
```
Comment (OOM-Killer) [inline scripts/wait_for_endpoint.sh]: Good call, going to replace that.

**Comment by OOM-Killer** [comment]
@Dieterbe You're right that from inside a container the `nc -z` problem does not exist because (I'm assuming) that `docker-proxy` isn't used within the container network. But in the case of the `scripts/wait_for_endpoint.sh` script I've generalized it and I'm reusing it in a second call site at `scripts/end2end_test.sh:20` now. That's outside a docker container and that's where it runs into that problem.
I could also keep these two scripts separate, then the one which is used to make `metrictank` wait for `cassandra:9042` would not need to be modified at all, I just thought it's nicer to generalize them because they basically do the same thing.

File: scripts/wait_for_endpoint.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    echo "waiting for $endpoint to become up..."
+
+    # connect and see if connection stays up.
+    # docker-proxy can listen to ports before the actual service is up,
+    # in which case it will accept and then close the connection again.
+    # this checks not only if the connect succeeds, but also if the 
+    # connection stays up for 3 seconds.
+    if [ $_using_busybox -eq 1 ]
+    then
+      busybox timeout -t 3 busybox nc $host $port
```
Comment (Dieterbe) [inline scripts/wait_for_endpoint.sh]: Ok

**Comment by Dieterbe** [comment]
ok sounds good.
just fix the basename and i'll merge.

**Comment by Dieterbe** [comment]
BTW, how crazy can we make this? could we potentially run larger stacks, more workload, several different use cases, ... ? will we run into restrictions like time or cpu or memory or number of containers?

**Comment by OOM-Killer** [comment]
@Dieterbe generally we can go pretty crazy. As far as I've seen the mem limit on my free Circle CI account is 4G per box. I don't remember what the time limit is, I think it might be 30min (at least for the SSH enabled ones). CPUs/I/O seem pretty solid based on my "feeling" when SSHing into it and building stuff there. So yeah, we can probably do quite a lot with that:)

**Comment by Dieterbe** [comment]
> Let's first merge this though. Then we can incrementally increase the crazy level.

yep. also my thoughts.

**Pull request merged**